File: src/main/java/io/vlingo/schemata/model/Organization.java
// Copyright © 2012-2018 Vaughn Vernon. All rights reserved.
//
// This Source Code Form is subject to the terms of the
// Mozilla Public License, v. 2.0. If a copy of the MPL
// was not distributed with this file, You can obtain
// one at https://mozilla.org/MPL/2.0/.

package io.vlingo.schemata.model;

import io.vlingo.actors.ActorInstantiator;
import io.vlingo.actors.Address;
import io.vlingo.actors.Definition;
import io.vlingo.actors.Stage;
import io.vlingo.common.Completes;
import io.vlingo.schemata.model.Id.OrganizationId;

public interface Organization {
  static String nameFrom(final OrganizationId organizationId) {
    return "O:"+organizationId.value;
  }

  static OrganizationId uniqueId() {
    return OrganizationId.unique();
  }

  static Completes<OrganizationState> with(
          final Stage stage,
          final String name,
          final String description) {
    return with(stage, uniqueId(), name, description);
  }

  static Completes<OrganizationState> with(
          final Stage stage,
          final OrganizationId organizationId,
          final String name,
          final String description) {
    final String actorName = nameFrom(organizationId);
    final Address address = stage.addressFactory().from(organizationId.value, actorName);
    final Definition definition = Definition.has(OrganizationEntity.class, new OrganizationInstantiator(organizationId), actorName);
    final Organization organization = stage.actorFor(Organization.class, definition, address);
    return organization.defineWith(name, description);
  }

  Completes<OrganizationState> defineWith(final String name, final String description);

  Completes<OrganizationState> describeAs(final String description);

  Completes<OrganizationState> redefineWith(final String name, final String description);

  Completes<OrganizationState> renameTo(final String name);

  static class OrganizationInstantiator implements ActorInstantiator<OrganizationEntity> {
    private final OrganizationId organizationId;

    public OrganizationInstantiator(final OrganizationId organizationId) {
      this.organizationId = organizationId;
    }

    @Override
    public OrganizationEntity instantiate() {
      return new OrganizationEntity(organizationId);
    }

    @Override
    public Class<OrganizationEntity> type() {
      return OrganizationEntity.class;
    }
  }
}

File: src/main/kotlin/io/vlingo/schemata/model/Organization.kt
// Copyright © 2012-2018 Vaughn Vernon. All rights reserved.
//
// This Source Code Form is subject to the terms of the
// Mozilla Public License, v. 2.0. If a copy of the MPL
// was not distributed with this file, You can obtain
// one at https://mozilla.org/MPL/2.0/.

package io.vlingo.schemata.model

import io.vlingo.lattice.model.sourcing.EventSourced
import io.vlingo.schemata.model.Events.OrganizationDefined
import io.vlingo.schemata.model.Events.OrganizationDescribed
import io.vlingo.schemata.model.Events.OrganizationRenamed
import io.vlingo.schemata.model.Id.OrganizationId

import java.util.function.BiConsumer

class Organization private constructor() : EventSourced() {
  private var state: State

  init {
    this.state = State()
  }

  fun describeAs(description: String) {
    apply(OrganizationDescribed(state.id, description))
  }

  fun renameTo(name: String) {
    apply(OrganizationRenamed(state.id, name))
  }

  internal fun applyDefined(event: OrganizationDefined) {
    state = State(OrganizationId.existing(event.organizationId), OrganizationId.existing(event.parentId), event.name, event.description)
  }

  internal fun applyDescribed(event: OrganizationDescribed) {
    state = state.withDescription(event.description)
  }

  internal fun applyRenamed(event: OrganizationRenamed) {
    state = state.withName(event.name)
  }

  inner class State @JvmOverloads internal constructor(val id: OrganizationId = OrganizationId.undefined(), val parentId: OrganizationId = OrganizationId.undefined(), val name: String = "", val description: String = "") {
    internal fun withDescription(description: String): State {
      return State(id, parentId, name, description)
    }

    internal fun withName(name: String): State {
      return State(id, parentId, name, description)
    }
  }

  companion object {
    init {
      val applyOrganizationDefinedFn = BiConsumer<Organization, OrganizationDefined>({ obj, event -> obj.applyDefined(event) })
      registerConsumer(Organization::class.java, OrganizationDefined::class.java, applyOrganizationDefinedFn)
      val applyOrganizationDescribedFn = BiConsumer<Organization, OrganizationDescribed>({ obj, event -> obj.applyDescribed(event) })
      registerConsumer(Organization::class.java, OrganizationDescribed::class.java, applyOrganizationDescribedFn)
      val applyOrganizationRenamedFn = BiConsumer<Organization, OrganizationRenamed>({ obj, event -> obj.applyRenamed(event) })
      registerConsumer(Organization::class.java, OrganizationRenamed::class.java, applyOrganizationRenamedFn)
    }

    fun defineWith(name: String, description: String): Organization {
      return defineWith(OrganizationId.undefined(), name, description)
    }

    fun defineWith(parentId: OrganizationId, name: String, description: String): Organization {
      assert(parentId.isDefined())
      assert(name.isNotEmpty())
      assert(description.isNotEmpty())
      val organization = Organization()
      organization.apply(OrganizationDefined(OrganizationId.unique(), parentId, name, description))
      return organization
    }
  }
}